### PR TITLE
#856: fixing crash because of bad array size

### LINKED ISSUE #856: Oversized arrays can cause runtime errors or crashes
The following program in Dafny successfully verifies:

```
method Crash() returns ()
{
  var big := 9223372036854775808;
  var a := new int[big];
  assert a.Length == big;
}
```

But the resulting C# program crashes when run.

```
public static void Crash()
{
  BigInteger _20_big;
  _20_big = new BigInteger(9223372036854775808UL);
  BigInteger[] _21_a;
  var _nw0 = new BigInteger[(int)(_20_big)];
  _21_a = _nw0;
}
```

Since the `big` integer cannot be cast from a `BigInteger` to an `int`, the result is an overflow runtime exception.

Similarly, in Java the `BigInteger` is cast to an `int` via `BigInteger.intValue()` which truncates the value to its lower 32 bits, and invalidates the assertion.

For this to be sound, Dafny either needs to enforce some upper bound on array sizes, or the target languages need to implement arrays with a collection type that can be indexed on BigIntegers directly.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #856.  Adding checks that casts to int from long or BigInteger are not out of range.

### CODE REVIEW

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java [hunk 1]
```diff
@@ -110,6 +110,49 @@ public static int toInt(BigInteger i) {
...
 
+    public static void outOfRange(String msg) {
+          //System.out.println(msg);
+          //System.exit(-1);
+          throw new dafny.DafnyHaltException(msg);
+    }
+
+    public static int toIntChecked(BigInteger i, String msg) {
+        int r = i.intValue();
+        if (!BigInteger.valueOf(r).equals(i) || (msg != null && r < 0)) {
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:121]: I don't understand the second disjunct. When would it ever be used in a situation that's not already caught by (the static verifier and) the first disjunct?

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java [hunk 2]
```diff
@@ -110,6 +110,49 @@ public static int toInt(BigInteger i) {
         return i.intValue();
     }
 
+    public static void outOfRange(String msg) {
+          //System.out.println(msg);
+          //System.exit(-1);
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:115]: Delete

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java [hunk 3]
```diff
@@ -110,6 +110,49 @@ public static int toInt(BigInteger i) {
...
+        if (!BigInteger.valueOf(r).equals(i) || (msg != null && r < 0)) {
+          msg = (msg != null ? msg : "value out of range for a 32-bit int") + ": " + i;
+          outOfRange(msg);
+        }
+        return r;
+    }
+
+    public static int toIntChecked(long i, String msg) {
+        int r = (int)i;
+        if (r != i || (msg != null && r < 0)) {
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:130]: See L121

File: Source/Dafny/Compiler-java.cs [hunk 4]
```diff
@@ -3483,20 +3483,24 @@ private class GenericArrayElementLvalue : ILvalue {
       }
     }
 
-    private void TrExprAsInt(Expression expr, TargetWriter wr, bool inLetExprBody) {
+    private void TrExprAsInt(Expression expr, TargetWriter wr, bool inLetExprBody, bool checkRange = false, string msg = null) {
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-java.cs:3486]: Are `checkRange` and `msg` independent parameters? If not, please write a contract about them. For example:

    Contract.Requires(!checkRange || msg != null);

If the relation between them turns out to be `checkRange == (msg != null)`, then I suggest eliminating the `checkRange` parameter and adding a descriptive comment for the method that explains that passing in `msg` as non-null gives rise to a checked conversion.

File: Source/Dafny/Compiler-Csharp.cs [hunk 5]
```diff
@@ -1309,10 +1309,17 @@ protected class ClassWriter : IClassWriter
...
+      wr.Write($"{GetHelperModuleName()}.ToIntChecked(");
+      TrExpr(expr, wr, inLetExprBody);
+      wr.Write($", \"{msg}\")");
+    }
+
     protected override void EmitNewArray(Type elmtType, Bpl.IToken tok, List<Expression> dimensions, bool mustInitialize, TargetWriter wr) {
       var wrs = EmitNewArray(elmtType, tok, dimensions.Count, mustInitialize, wr);
       for (int i = 0; i < wrs.Count; i++) {
-        TrExpr(dimensions[i], wrs[i], inLetExprBody: false);
+        TrExprAsInt(dimensions[i], wrs[i], inLetExprBody: false, true, "C# arrays may not be larger than the max 32-bit integer");
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-Csharp.cs:1322]: I like the way you resolved this!

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Source/Dafny/Compiler-Csharp.cs (hunk 5 above)
Comment (davidcok) [reply Source/Dafny/Compiler-Csharp.cs:1322]: :-)

File: Source/Dafny/Compiler-java.cs (hunk 4 above)
Comment (davidcok) [reply Source/Dafny/Compiler-java.cs:3486]: They are independent, except that if checkRange == false then msg is ignored.
Code cleaned up and commented to indicate this

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java (hunk 2 above)
Comment (davidcok) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:115]: Done

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java (hunk 1 above)
Comment (davidcok) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:121]: Obsolete code -- removed

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java (hunk 3 above)
Comment (davidcok) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Helpers.java:130]: Obsolete code - removed

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**